File: assets/js/user_socket.js
// Bring in Phoenix channels client library:
import {Socket} from "phoenix"

// And connect to the path in "lib/quic_web/endpoint.ex". We pass the
// token for authentication. Read below how it should be used.

// When you connect, you'll often need to authenticate the client.
// For example, imagine you have an authentication plug, `MyAuth`,
// which authenticates the session and assigns a `:current_user`.
// If the current user exists you can assign the user's token in
// the connection for use in the layout.
//
// In your "lib/quic_web/router.ex":
//
//     pipeline :browser do
//       ...
//       plug MyAuth
//       plug :put_user_token
//     end
//
//     defp put_user_token(conn, _) do
//       if current_user = conn.assigns[:current_user] do
//         token = Phoenix.Token.sign(conn, "user socket", current_user.id)
//         assign(conn, :user_token, token)
//       else
//         conn
//       end
//     end
//
// Now you need to pass this token to JavaScript. You can do so
// inside a script tag in "lib/quic_web/templates/layout/app.html.heex":
//
//     <script>window.userToken = "<%= assigns[:user_token] %>";</script>
//
// You will need to verify the user token in the "connect/3" function
// in "lib/quic_web/channels/user_socket.ex":
//
//     def connect(%{"token" => token}, socket, _connect_info) do
//       # max_age: 1209600 is equivalent to two weeks in seconds
//       case Phoenix.Token.verify(socket, "user socket", token, max_age: 1_209_600) do
//         {:ok, user_id} ->
//           {:ok, assign(socket, :user, user_id)}
//
//         {:error, reason} ->
//           :error
//       end
//     end
//
// Finally, connect to the socket:
// socket.connect()

let channel;

// Function for a User to join the channel of a Session
function joinChannel(session_code, username, isMonitor, session_id = null) {
  let socket = new Socket("/socket", {params: {session: session_code, username: username, isMonitor: isMonitor }})
  socket.connect()
  window.channel_socket = socket;

  if (session_id) {
    channel = socket.channel(
      "session:" + session_code, 
      {"username": username, "isMonitor": isMonitor, "session_id": session_id}
    )
  } else {
    channel = socket.channel(
      "session:" + session_code, 
      {"username": username, "isMonitor": isMonitor}
    )
  }

  channel.join()
    .receive("ok", () => {window.channel = channel})
    .receive("error", () => socket.disconnect());
}

File: lib/quic_web/channels/user_socket.ex
defmodule QuicWeb.UserSocket do
  use Phoenix.Socket

  require Logger

  # A Socket handler
  #
  # It's possible to control the websocket connection and
  # assign values that can be accessed by your channel topics.

  ## Channels

  channel "session:*", QuicWeb.SessionChannel

  # Socket params are passed from the client and can
  # be used to verify and authenticate a user. After
  # verification, you can put default assigns into
  # the socket that will be set for all channels, ie
  #
  #     {:ok, assign(socket, :user_id, verified_user_id)}
  #
  # To deny connection, return `:error` or `{:error, term}`. To control the
  # response the client receives in that case, [define an error handler in the
  # websocket
  # configuration](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html#socket/3-websocket-configuration).
  #
  # See `Phoenix.Token` documentation for examples in
  # performing token verification on connect.
  @impl true
  def connect(%{"username" => username, "session" => code, "isMonitor" => isMonitor} = _params, socket, _connect_info) do
    {:ok, assign(socket, username: username, isMonitor: isMonitor, session_code: code)}
  end


  # Socket id's are topics that allow you to identify all sockets for a given user:
  #     def id(socket), do: "user_socket:#{socket.assigns.user_id}"
  # Would allow you to broadcast a "disconnect" event and terminate
  # all active sockets and channels for a given user:
  #     Elixir.QuicWeb.Endpoint.broadcast("user_socket:#{user.id}", "disconnect", %{})
  # Returning `nil` makes this socket anonymous.
  @impl true
  def id(socket) do
    if socket.assigns.isMonitor === true do
      Logger.error("user_socket.ex says: is IS a monitor")
      "session:#{socket.assigns.session_code}:monitor:#{socket.assigns.username}"
    else
      Logger.error("user_socket.ex says: is NOT a monitor")
      "session:#{socket.assigns.session_code}:participant:#{socket.assigns.username}"
    end
  end
end
